% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/read_vcfs_as_granges.R
\name{read_vcfs_as_granges}
\alias{read_vcfs_as_granges}
\title{Read VCF files into a GRangesList}
\usage{
read_vcfs_as_granges(vcf_files, sample_names, genome = "-", style = "UCSC")
}
\arguments{
\item{vcf_files}{Character vector of vcf file names}

\item{sample_names}{Character vector of sample names}

\item{genome}{A character or Seqinfo object}

\item{style}{The naming standard to use for the GRanges. (default = "UCSC")}
}
\value{
A GRangesList containing the GRanges obtained from vcf_files
}
\description{
This function reads Variant Call Format (VCF) files into a GRanges object
and combines them in a GRangesList.  In addition to loading the files, this
function applies a seqlevel style to the GRanges objects.  The default
seqlevel style is "UCSC".
}
\examples{
# The example data set consists of three colon samples, three intestine
# samples and three liver samples.  So, to map each file to its appropriate
# sample name, we create a vector containing the sample names:
sample_names <- c("colon1", "colon2", "colon3",
                    "intestine1", "intestine2", "intestine3",
                    "liver1", "liver2", "liver3")

# We assemble a list of files we want to load.  These files match the
# sample names defined above.
vcf_files <- list.files(system.file("extdata", 
                                    package="MutationalPatterns"),
                                    pattern = ".vcf", full.names = TRUE)

# This function loads the files as GRanges objects
vcfs <- read_vcfs_as_granges(vcf_files, sample_names, genome = "hg19")

}

